pkg/controller/rbac/application/managed.go

```diff
@@ -0,0 +1,195 @@
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/password"
+	"github.com/crossplane/crossplane-runtime/pkg/ratelimiter"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	"github.com/crossplane/provider-azure/apis/rbac/v1alpha1"
+	azure "github.com/crossplane/provider-azure/pkg/clients"
+	azureclients "github.com/crossplane/provider-azure/pkg/clients"
+)
+
+// Error strings.
+const (
+	errNotApplication    = "managed resource is not an Application"
+	errCreateApplication = "cannot create Application"
+	errUpdateApplication = "cannot update Application"
+	errGetApplication    = "cannot get Application"
+	errDeleteApplication = "cannot delete Application"
+)
+
+// Setup adds a controller that reconciles Application.
+func Setup(mgr ctrl.Manager, l logging.Logger, rl workqueue.RateLimiter, poll time.Duration) error {
+	name := managed.ControllerName(v1alpha1.ApplicationKind)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithOptions(controller.Options{
+			RateLimiter: ratelimiter.NewDefaultManagedRateLimiter(rl),
+		}).
+		For(&v1alpha1.Application{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1alpha1.ApplicationGroupVersionKind),
+			// Override default initializers in case to remove NewNameAsExternalName Initializer
+			managed.WithInitializers(),
+			managed.WithConnectionPublishers(),
+			managed.WithExternalConnecter(&connecter{client: mgr.GetClient()}),
+			managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+			managed.WithPollInterval(poll),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type connecter struct {
+	client client.Client
+}
+
+func (c *connecter) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	creds, _, err := azureclients.GetAuthInfo(ctx, c.client, mg)
+	if err != nil {
+		return nil, err
+	}
+	cfg, err := adal.NewOAuthConfig(creds[azure.CredentialsKeyActiveDirectoryEndpointURL], creds[azure.CredentialsKeyTenantID])
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot create OAuth configuration")
+	}
+	token, err := adal.NewServicePrincipalToken(*cfg,
+		creds[azure.CredentialsKeyClientID],
+		creds[azure.CredentialsKeyClientSecret],
+		creds[azure.CredentialsKeyActiveDirectoryGraphResourceID])
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot create service principal token")
+	}
+	if err := token.Refresh(); err != nil {
+		return nil, errors.Wrap(err, "cannot refresh service principal token")
+	}
+	ta := autorest.NewBearerAuthorizer(token)
+	ac := graphrbac.NewApplicationsClient(creds[azure.CredentialsKeyTenantID])
+	ac.Authorizer = ta
```

Comment:
Why don't we make use of the `autorest.Authorizer` returned from `azureclients.GetAuthInfo`? Does it not work in this case?:
```suggestion
	creds, auth, err := azureclients.GetAuthInfo(ctx, c.client, mg)
	if err != nil {
		return nil, err
	}
	ac := graphrbac.NewApplicationsClient(creds[azure.CredentialsKeyTenantID])
	ac.Authorizer = auth
```